File: Javascript/rotate_linked_list.js
function Node(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : right)
}
// Time complexity: O(n)
// Space complexity: O(1)
var rotateLinkedList = (head, k) => {
    var len = 0
    var cur = head
    var tail = null
    while (cur) {
        len++
        tail = cur
        cur = cur.next
    }
    k = len - k % len - 1
    cur = head
    while (k) {
        cur = cur.next
        k--
    }
    tail.next = head
    head = cur.next
    cur.next = null
    return head
}

head = new Node(1)
head.next = new Node(2)
head.next.next = new Node(3)
head.next.next.next = new Node(4)
head.next.next.next.next = new Node(5)

console.log(rotateLinkedList(head, 1))

File: Python/rotate_linked_list.py
class Node(object):
    def __init__(self, val, next=None):
        self.val = val
        self.next = next

    def __repr__(self):
        result = ''
        head = self
        while head:
            result += str(head.val) + ' '
            head = head.next

        return result

class Solution():
    def __init__(self):
        self.name = 'S->13'

    # Time complexity: O(n)
    # Space complexity: O(1)
    def rotateLinkedList(self, head, k):
        cur = head
        tail = None
        length = 0
        while cur:
            length += 1
            tail = cur
            cur = cur.next
        k = length - k % length - 1
        cur = head
        while k:
            cur = cur.next
            k -= 1
        tail.next = head
        head = cur.next
        cur.next = None

        return head

head = Node(1, Node(2, Node(3, Node(4, Node(5)))))
print(Solution().rotateLinkedList(head, 4))
